lib/Config_option.ml

```diff
@@ -188,46 +188,91 @@ module Make (C : CONFIG) = struct
     store := Pack opt :: !store ;
     opt
 
-  type removed_value = {name: string; version: string; msg: string}
-
-  let removed_value ~name ~version ~msg = {name; version; msg}
-
-  let removed_values ~names ~version ~msg =
-    List.map names ~f:(fun name -> removed_value ~name ~version ~msg)
-
-  let add_removed_values values conv =
-    let parse s =
-      match List.find values ~f:(fun {name; _} -> String.equal name s) with
-      | Some {name; version; msg} ->
-          Format.kasprintf
-            (fun s -> Error (`Msg s))
-            "value `%s` has been removed in version %s. %s" name version msg
-      | None -> Arg.conv_parser conv s
-    in
-    Arg.conv (parse, Arg.conv_printer conv)
+  module Value = struct
+    module Ok = struct
+      type 'a t = string * 'a * string * [`Valid | `Deprecated of deprecated]
+
+      let valid ~name ~doc value = (name, value, doc, `Valid)
+
+      let deprecated ~name ~doc ~deprecated value =
+        (name, value, doc, `Deprecated deprecated)
+
+      let pp_deprecated s ppf {dmsg= msg; dversion= v} =
+        Format.fprintf ppf "Value `%s` is deprecated since version %s. %s" s
+          v msg
+
+      let pp_deprecated_with_name ~opt ~val_ ppf {dmsg= msg; dversion= v} =
+        Format.fprintf ppf
+          "Value `%s` of option `%s` is deprecated since version %s. %s" val_
+          opt v msg
+
+      let status_doc s ppf = function
+        | `Valid -> ()
+        | `Deprecated x ->
+            Format.fprintf ppf " Warning: %a" (pp_deprecated s) x
+
+      let warn_if_deprecated conf opt (s, _, _, status) =
+        match status with
+        | `Valid -> ()
+        | `Deprecated d ->
+            C.warn conf "%a" (pp_deprecated_with_name ~opt ~val_:s) d
+    end
+
+    module Removed = struct
+      type t = {name: string; version: string; msg: string}
+
+      let mk ~name ~version ~msg = {name; version; msg}
+
+      let mk_list ~names ~version ~msg =
+        List.map names ~f:(fun name -> mk ~name ~version ~msg)
+
+      let add_parse_errors values conv =
+        let parse s =
+          match
+            List.find values ~f:(fun {name; _} -> String.equal name s)
+          with
+          | Some {name; version; msg} ->
+              Format.kasprintf
+                (fun s -> Error (`Msg s))
+                "value `%s` has been removed in version %s. %s" name version
```

Comment:
This doesn't say "of option '%s'" like the deprecated message.

Reply:
This is not necessary as the error message issued when parsing the value is preceded by the option name:
```
$ ocamlformat a.ml --break-string-literals=wrap
ocamlformat: option `--break-string-literals': value `wrap` has been removed
             in version 0.12. It has been replaced by the new default `auto`
             value, which breaks lines at newlines and wraps string literals
             at the margin.
```

Reply:
Indeed. It would be nice to use the same template for deprecation messages.

Reply:
done:
```
Warning: option `break-string-literals`: value `never` is deprecated since version V. Deprecation message.
```